### PR TITLE
kubelet/container: Refactor RunContainerOptions.

### PR DESCRIPTION
**Pull request by yifan-gu**
Make Envs, Mounts, PortMappings more generic.
Also add default name for PortMapping if it's not specified.

Also addressed #7710 with `kubelet.makePortMappings()`

Will add tests for it and run e2e.

/cc @vmarmol @dchen1107 @yujuhong @thockin

### CODE REVIEW

**Comment by yujuhong** [comment]
LGTM overall with one small nit.

File: pkg/kubelet/container/runtime.go [hunk 1]
```diff
@@ -185,14 +185,38 @@ type Image struct {
...
+	// The host IP.
+	HostIP string
+}
+
 // RunContainerOptions specify the options which are necessary for running containers
 type RunContainerOptions struct {
-	// The environment variables, they are in the form of 'key=value'.
-	Envs []string
+	// The environment variables map.
+	Envs map[string]string
```
Comment (pmorie) [inline pkg/kubelet/container/runtime.go]: Not sure if it matters, but ordering on envs is going to be affected by this.

Comment (vmarmol) [inline pkg/kubelet/container/runtime.go]: Hmm some people have env vars that depend on other env vars so the order is probably important.

Comment (vmarmol) [inline pkg/kubelet/container/runtime.go]: Let's revert to using the slice.

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -473,6 +473,65 @@ func (dm *DockerManager) GetPodInfraContainer(pod kubecontainer.Pod) (kubecontai
 	return kubecontainer.Container{}, fmt.Errorf("unable to find pod infra container for pod %v", pod.ID)
 }
 
+// makeEnvList converts the map to a list of strings, in the form of
+// '<key>=<value>', which can be understood by docker.
+func makeEnvList(envs map[string]string) (result []string) {
+	for k, v := range envs {
+		result = append(result, k, v)
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/manager.go]: I don't think this inserts with an "=" in between? But we may remove this if we change to a list.

Comment (pmorie) [inline pkg/kubelet/dockertools/manager.go]: You're right @vmarmol re append, but it doesn't matter if we revert to
using the list.

**Comment by yifan-gu** [comment]
Comments address. I also changed the env map to env list to take care of the ordering as mentioned by @pmorie , thank you!

e2e is passing
```
Ran 39 of 45 Specs in 1639.770 seconds
SUCCESS! -- 39 Passed | 0 Failed | 1 Pending | 5 Skipped 2015/05/13 10:57:15 Running: teardown
```

**Comment by vmarmol** [comment]
LGTM, thanks @yifan-gu!

**Comment by yujuhong** [comment]
LGTM again. It'd be nice to add a comment to indicate that the order of the env variables should be respected :)

**Pull request merged**